File: src/main/java/no/nav/personopplysninger/features/sts/STSConsumer.java
package no.nav.personopplysninger.features.sts;

import no.nav.personopplysninger.features.ConsumerException;

import javax.ws.rs.ProcessingException;
import javax.ws.rs.client.Client;
import javax.ws.rs.client.Invocation;
import javax.ws.rs.core.Response;
import java.net.URI;

public class STSConsumer {

    private Client client;
    private URI endpoint;

    public STSConsumer(Client client, URI endpoint) {
        this.client = client;
        this.endpoint = endpoint;
    }

    public TokenDto getToken() {
        Invocation.Builder request = buildSTSRequest();
        return getToken(request);
    }

    private Invocation.Builder getBuilder(String path) {
        return client.target(endpoint)
                .path(path)
                .queryParam("grant_type", "client_credentials")
                .queryParam("scope", "openid")
                .request();
    }

    private Invocation.Builder buildSTSRequest() {
        return getBuilder("");
    }

    private TokenDto getToken(Invocation.Builder request) {
        try (Response response = request.get()) {
            return readEntity(TokenDto.class, response);
        } catch (Exception e) {
            String msg = "Forsøkte å hente STSToken. endpoint=[" + endpoint + "].";
            throw new ConsumerException(msg, e);
        }
    }


    private <T> T readEntity(Class<T> responsklasse, Response response) {
        try {
            return response.readEntity(responsklasse);
        } catch (ProcessingException e) {
            throw new RuntimeException("Prosesseringsfeil på responsobjekt. Responsklasse: " + e.getStackTrace() + " " + responsklasse.getName(), e);
        } catch (IllegalStateException e) {
            throw new RuntimeException("Ulovlig tilstand på responsobjekt. Responsklasse: " + responsklasse.getName(), e);
        } catch (Exception e) {
            throw new RuntimeException("Uventet feil på responsobjektet. Responsklasse: " + responsklasse.getName(), e);
        }
    }
}

File: src/main/kotlin/no/nav/personopplysninger/oppslag/sts/STSConsumer.kt
package no.nav.personopplysninger.oppslag.sts

import com.fasterxml.jackson.module.kotlin.readValue
import no.nav.personopplysninger.util.ConsumerException
import no.nav.personopplysninger.util.JsonDeserialize
import no.nav.personopplysninger.util.consumerErrorMessage
import java.net.URI
import javax.ws.rs.client.Client
import javax.ws.rs.client.Invocation
import javax.ws.rs.core.Response

class STSConsumer(private val client: Client, private val endpoint: URI) {

    val token: TokenDto
        get() {
            val request = buildSTSRequest()
            return getToken(request)
        }

    private fun getBuilder(): Invocation.Builder {
        return client.target(endpoint)
            .queryParam("grant_type", "client_credentials")
            .queryParam("scope", "openid")
            .request()
    }

    private fun buildSTSRequest(): Invocation.Builder {
        return getBuilder()
    }

    private fun getToken(request: Invocation.Builder): TokenDto {
        request.get().use { response ->
            val responseBody = response.readEntity(String::class.java)
            if (Response.Status.Family.SUCCESSFUL != response.statusInfo.family) {
                throw ConsumerException(consumerErrorMessage(endpoint, response.status, responseBody))
            }
            return JsonDeserialize.objectMapper.readValue(responseBody)
        }
    }
}
